% XeLaTeX can use any Mac OS X font. See the setromanfont command below.
% Input to XeLaTeX is full Unicode, so Unicode characters can be typed directly into the source.

% The next lines tell TeXShop to typeset with xelatex, and to open and save the source with Unicode encoding.

%!TEX TS-program = xelatex
%!TEX encoding = UTF-8 Unicode

\documentclass[12pt]{article}
\usepackage{geometry}                % See geometry.pdf to learn the layout options. There are lots.
\geometry{letterpaper}                   % ... or a4paper or a5paper or ... 
%\geometry{landscape}                % Activate for for rotated page geometry
%\usepackage[parfill]{parskip}    % Activate to begin paragraphs with an empty line rather than an indent
\usepackage{graphicx}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{hyperref}

% Will Robertson's fontspec.sty can be used to simplify font choices.
% To experiment, open /Applications/Font Book to examine the fonts provided on Mac OS X,
% and change "Hoefler Text" to any of these choices.

\usepackage{fontspec,xltxtra,xunicode}
\defaultfontfeatures{Mapping=tex-text}
\setromanfont[Mapping=tex-text]{Hoefler Text}
\setsansfont[Scale=MatchLowercase,Mapping=tex-text]{Gill Sans}
\setmonofont[Scale=MatchLowercase]{Andale Mono}
\DeclareMathOperator*{\argmax}{argmax}

\title{Network analysis of heterogenous cell populations across different perturbations with a mixture of Nested Effects Models}
\author{Martin Pirkl \& Niko Beerenwinkel}
%\date{}                                           % Activate to display a given date or no date

\begin{document}
\maketitle

\begin{abstract}
New technologies allow for the elaborate measurement of different traits of single cells. These data promise the opportunity to elucidate causal intra-cellular mechanisms in unprecedented detail. These insight do not only extend our general knowledge of cells, but give evidence of why they cease to function properly or are de-regulated. That de-regulation can lead to life threatening diseases like cancer. The battle against those diseases will benefit from our understanding of cellular function on the single cell level.

However, cell populations can be very heterogeneous and consist of several sub populations. These sub populations do not necessarily share the same signalling pathway structure. Under that assumption we us a mixture of Nested Effects Models to identify not only the different sub populations, but also the network structure of signalling pathways for each one.

We show that our method performs well in silico under reasonable assumptions and apply it to two different pooled CrispR screens.

The mixture Nested Effects Model (M\&NEM) is available as an R-package at \url{https://bitbucket.org/MartinFXP/mnem/}.
\end{abstract}

\section{Introduction}
Understanding diseases like cancer on a molecular level is crucial for the improvement and development of treatment therapies. For example intra-tumor heterogeneity is an important factor when it comes to cancer treatment (Sun, et al. 2015, Prasetyanti, et al. 2017). These treatments assume cancer to be homogeneous across cells. However, if different celltypes are resistent to different treatments, the success of current treatment strategies are limited.

A key component of the fight against illness is our knowledge of signalling pathways and how they are wired in healthy and unhealthy cells. De-regulation of pathways in unhealthy cells is not uncommon (Mao,  et al. 2012, Giancotti, et al. 2014). To study this de-regulation, different mathematical methods have been developed.

Nested Effects Models (NEM, Markowetz, et al. 2005, 2007) infers pathways from perturbation data. In each experiment one protein in the pathway is knocked down and gene expression is measured. If the expression of one gene changes during the knock-down compared to the unperturbed control, the knock-down has an effect on the gene aka the gene responds to the knock-down. If the genes responding to the knock-down of protein B are a noisy subset of the genes responding to the knock-down of protein A, we place A upstream of B in the pathway and a causal edge leads from A to B. A causes B. NEMs have been successfully applied to different biological data sets to infer the network of signalling pathways (Markowetz, et al. 2005, Froehlich, et al. 2009, McNeil, et al. 2015). Several extensions of NEM have been developed, e.g. to account for hidden variables (Sadeh, et al. 2013). Epistatic Nested Effect Models (EpiNEMs, Pirkl, et al. 2017) systematically infer epistasis from double knock-down screens. Boolean Nested Effect Models make use of arbitrary combinations of knock-downs and knock-ins per experiment to infer a full boolean network and additionally integrate literature knowledge in the inference. Dynamic Nested Effect Models (D-NEM, Anchang, et al. 2009, Froehlich, et al. 2011) infer the flow speed of the signal from time series data, while Hidden Markow Nested Effect Models (HM-NEM, Wang, et al. 2015) model the evolution of the network itself during this time course.

The arrival of single cell technology provides new opportunities to improve resolution and account for heterogeneity in a population of cells. Pooled CrispR screens make it possible to have gene expression measurements for thousands of cells with each cell having been the target of a CrispR modification, i.e. a knock-down (Dixit, et al. 2016, Datlinger, et al. 2017).

We developed a mixture model, which infers different sub-population of cells across knock-downs. For each sub-population, we reconstruct a network of the pathway of the genes targeted by the knock-downs. This is similar to a Gaussian mixture model (GMM) except for data points drawn from a Gaussian distribution we have single cells and their transcriptome read-out and instead of mean and standard deviations for several components we have different pathways (Toy example in Figure~\ref{fig:toy}). As with GMMs the cells are not hard clustered but soft such as each cell has a certain probability of being generated by each component.

We show that Mixture Nested Effect Models (M\&NEM) work in silico and apply our novel method to two different pooled CrispR screens.

\begin{figure}
\begin{center}
\includegraphics[width=1\textwidth]{cells.pdf}
\includegraphics[width=0.4\textwidth]{comps.pdf}
\end{center}
\caption{\textbf{M\&NEM toy example} \textbf{(A)} Two dimensional projection of unlabelled cells. \textbf{(B)} Cells labelled for their known knock-down(s). \textbf{(C)} The unknown underlying mixture components. \textbf{(D)} A potential set of components responsible for the heterogeneous cell population.}\label{fig:toy}
\end{figure}


\section{Method}
\subsection{Nested Effects Model}
A Nested Effects Model is parametrised as an adjacency matrix $\Phi \in M_{n,n}(\left\{0,1\right\})$ for the directed acyclic graph (DAG) representation of the signalling graph and an adjacency matrix $\Theta \in M_{n,m}(\left\{0,1\right\})$ for the E-gene attachments. However, $\Theta$ has S-genes exclusively in the rows and E-genes in the columns. Each column of $\Theta$ has only one non-zero entry, because each E-gene can have only one parent. We calculate the expected E-genes pattern for a given model $\Phi, \Theta$ by the matrix product
\begin{equation*}
F = \Phi\Theta \in M_{n,m}(\left\{0,1\right\}).
\end{equation*}
Let $D = (d_{ij}) \in M_{m,l}(\mathbb{R})$ be the raw data matrix and  $R = (r_{ij}) \in M_{m,l}(\mathbb{R})$ the log ratio matrix with perturbed cells as independent columns and observed genes as rows. We write the entries of $R$ as
\begin{equation*}
r_{ij} = \mathrm{log}\left( \frac{P(d_{ij}|e_{ij} = 1)}{P(d_{ij}|e_{ij} = 0)} \right).
\end{equation*}
with $e_{ij}$ the state of E-gene $i$ in cell $j$. As in Tresch \& Markowetz 2008 we write the log likelihood ratio of a give model $\Phi, \Theta$ and the null model $N$ , which predict no effects, as
\begin{align*}
\mathrm{log}P(D|\Phi,\Theta) - \mathrm{log}P(D|N) &= \mathrm{tr}(FR)\nonumber\\
&\Leftrightarrow\nonumber\\
\mathrm{log}P(D|\Phi, \Theta) &= \mathrm{tr}(FR) + \mathrm{log}P(D|N) = \mathrm{tr}(FR) + C
\end{align*}
with $\mathrm{tr}$ as the trace function of a quadratic matrix. We assume $l=n$. If the data contains replicates for each knock-down, they have to be summarised beforehand.

\subsection{Mixture Nested Effects Model}
The model parameters for a mixture of $K$ components are
\begin{equation*}
(\Phi, \Theta ) = ( \Phi_k, \Theta_k )_{k = 1,\dots,K} 
\end{equation*}
with $k = 1, \dots, K$. Given a component $(\Phi_k, \Theta_k)$ we calculate the expected knock-down profiles for all single perturbations the same as the expected E-gene profiles by
\begin{equation*}
F^k = ( f_{ij}^k ) = \Phi_k \Theta_k \in M_{n,m}(\left\{0,1\right\}),
\end{equation*}
with $f^k_{ij}$ as the expected value of E-gene $j$ during the perturbation of S-gene $i$.
$\Omega = ( \omega_{ij} ) \in M_{n,l}(\left\{0,1\right\})$ is the perturbation matrix of the cells with $\omega_{ij} = 1$, if cell $j$ has undergone perturbation $i$ and $0$ otherwise. We calculate expected cell profiles for each cell with single or multiple perturbations with
\begin{equation}\label{eq:Pk}
P^k = (p^k_{ij}) = (F^k)^T \Omega \in M_{m,l}(\left\{0,1\right\}.
\end{equation}
Since the entries of $P^k$ are the expected binary effects, all non-zero entries in $P^k$ have to be normalised to $1$. Analogous to Tresch \& Markowetz (2008) we score the log odds ratio profile of each cell given component $k$ by
\begin{equation}\label{eq:Lk}
L^k = (l^k_{ij}) = (P^k)^T R \in M_{l,l}(\mathbb{R})
\end{equation}
and calculate the derivation of the likelihood of component $k$ by
\begin{align*}
\mathrm{log}P(D|P^k) - \mathrm{log}P(D|N) &= \mathrm{tr}(L^k)\nonumber\\
&\Leftrightarrow\nonumber\\
\mathrm{log}P(D|\Phi_k, \Theta_k) &= \mathrm{tr}(L^k) + \mathrm{log}P(D|N) = \mathrm{tr}(L^k) + C
\end{align*}
with N as the null model predicting no effects and $\mathrm{tr}$ as the trace function of a quadratic matrix.

Let $Z \in M_{K,l}(\left\{0,1\right\})$ be a matrix with $z_{ki} = 1$, if cell $i$ belongs to component $k$. Each column of $Z$ has exactly one non-zero entry. The distribution of $Z$ is defined by the mixing coefficients $\pi_k$ with
\begin{equation*}
P(z_{ki} = 1) = \pi_k \in [0,1] ~\forall i \in \left\{1,\dots,l\right\}
\end{equation*}
with $\pi = (\pi_1,\dots,\pi_K)$ and $\sum\limits_k \pi_k = 1$.

\paragraph{Log likelihood of the mixture}
We formulate a likelihood, which we use for model optimisation. We proceed similarly to the formulation for a single mixture component.
\begin{align}
&\mathrm{log}P(D| \Phi,\Theta) - \mathrm{log}P(D|N) = \mathrm{log} \prod\limits_{j=1}^l \frac{P(d_{j}|\Phi,\Theta)}{P(d_{j}|N)} = \mathrm{log} \prod\limits_{j=1}^l \frac{\sum\limits_Z P(Z) P(d_{j}|\Phi,\Theta, Z)}{P(d_{j}|N)}\nonumber\\
&\stackrel{*}{=} \mathrm{log} \prod\limits_{j=1}^l \frac{\sum\limits_{k=1}^K P(z_{kj} = 1) P(d_{j}|\Phi,\Theta, z_{kj} = 1)}{P(d_{j}|N)} = \mathrm{log} \prod\limits_{j=1}^l \sum\limits_{k=1}^K \pi_k \prod\limits_{i=1}^m \frac{P(d_{ij}|\Phi,\Theta, z_{kj} = 1)}{P(d_{ij}|N)}\nonumber\\
&\stackrel{\eqref{eq:Pk}}{=} \mathrm{log} \prod\limits_{j=1}^l \sum\limits_{k=1}^K \pi_k \prod\limits_{i=1}^m \frac{P(d_{ij}|e_{ij} = p^k_{ij})}{P(d_{ij}|e_{ij} = 0)} = \mathrm{log} \prod\limits_{j=1}^l \sum\limits_{k=1}^K \pi_k \prod\limits_{i=1}^m 
\begin{cases}
\mathrm{exp}(r_{ij}) &\text{ if } p_{ji}^k = 1\\
1 = \mathrm{exp}(0) &\text{ if } p_{ji}^k = 0\\
\end{cases}
\nonumber\\
&= \mathrm{log} \prod\limits_{j=1}^l \sum\limits_{k=1}^K \pi_k \prod\limits_{i=1}^m \mathrm{exp}(p_{ji}^k r_{ij}) = \mathrm{log} \prod\limits_{j=1}^l \sum\limits_{k=1}^K \pi_k \mathrm{exp}\left(\sum\limits_{i=1}^m p_{ji}^kr_{ij}\right)\nonumber\\
&\stackrel{\eqref{eq:Lk}}{=} \mathrm{log} \prod\limits_{j=1}^l \sum\limits_{k=1}^K \pi_k \mathrm{exp}(l^k_{jj}) = \sum\limits_{j=1}^l \mathrm{log} \sum\limits_{k=1}^K \pi_k \mathrm{exp}(l^k_{jj}) = \mathrm{tr}\left( \mathrm{log} \sum\limits_{k=1}^K \pi_k \mathrm{exp}(L^k) \right).
\end{align}
$*$ holds, because
\[z_{kj} = 1 \Rightarrow z_{ij} = 0 ~\forall i \neq k.\]
Analogous to the previous likelihood formulation,
\begin{align}\label{eq:logl}
\mathrm{log}P(D|\Phi,\Theta) = \mathrm{tr}\left( \mathrm{log} \sum\limits_{k=1}^K \pi_k \mathrm{exp}(L^k) \right) + C.
\end{align}

\subsection{Inference with a nested Expectation maximisation algorithm}
We compute the responsibility of component $k$ for cell $i$.
\begin{align}\label{eq:gammaki}
\gamma(z_{ki}) &= P(z_{ki} = 1|d_i) = \frac{P(z_{ki} = 1)P(d_i|z_{ki} = 1)}{\sum\limits_{j=1}^KP(z_{ji} = 1)P(d_i|z_{ji} = 1)} = \frac{P(z_{ki} = 1)\frac{P(d_i|z_{ki} = 1)}{P(d_i|N)}}{\sum\limits_{j=1}^KP(z_{ji} = 1)\frac{P(d_i|z_{ji} = 1)}{P(d_i|N)}}\nonumber\\
&=   \frac{\pi_k\prod\limits_{t=1}^m\frac{P(d_{ti}|e_{ti} = p_{it}^k)}{P(d_{ti}|e_{ti} = 0)}}{\sum\limits_{j=1}^K\pi_j\prod\limits_{t=1}^m\frac{P(d_{ti}|e_{ti} = p_{it}^j)}{P(d_{ti}|e_{ti} = 0)}} =   \frac{\pi_k\prod\limits_{t=1}^m \mathrm{exp}(p_{it}^kr_{ti})}{\sum\limits_{j=1}^K\pi_j\prod\limits_{t=1}^m\mathrm{exp}(p_{it}^j r_{ti})} = \frac{\pi_k\mathrm{exp}\left(\sum\limits_{t=1}^m p_{it}^k r_{ti}\right)}{\sum\limits_{j=1}^K\pi_j\mathrm{exp}\left(\sum\limits_{t=1}^mp_{it}^j r_{ti}\right)}\nonumber\\
&= \frac{\pi_k \mathrm{exp}\left(l_{ii}^k\right)}{\sum\limits_{j=1}^K\pi_j \mathrm{exp}\left(l_{ii}^j\right)},
\end{align}
which we summarise in
\begin{equation*}
\Gamma = (\gamma_{ki}) = (\gamma(z_{ki})) \in M_{K,l} (\left(0,1\right)).
\end{equation*}
 
\paragraph{M step}
Given $\Gamma$, we optimise each component $(\Phi_k, \Theta_k)$. We account for the responsibilities by a weighted log ratio matrix $R_k$, which we define by
\begin{equation*}
R_k = ( r_{ij} \gamma_{kj} ),% ~\forall i \in \left\{1,\dots,m\right\},
\end{equation*}
 We use the likelihood defined in Tresch \& Markowetz (2008) to find a new optimum $\left(\Phi_k^{new}, \Theta_k^{new}\right)$.

\paragraph{E step}
Let $\pi, (\Phi, \Theta)$ be the current parametrisation of our mixture. We calculate $L_k$ from equ.~\eqref{eq:Lk} with $R_k$ substituted for $R$ for every component $k$ and subsequently the responsibilities (equ.~\eqref{eq:gammaki}) and the log likelihood (equ.~\eqref{eq:logl}). We also update $\pi$ with
\begin{equation*}
\pi_k = \frac{\sum\limits_{i=1}^l \gamma_{ki}}{\sum\limits_{j=1}^K\sum\limits_{i=1}^l \gamma_{ji}}.
\end{equation*}
In a nested EM with a fixed $\Phi = (\Phi_k)$ we iteratively estimate $\Theta = (\Theta_k)$ and update $\pi$ given $\Phi$ and $\Gamma$ (nested M step) and re-calculate the responsibilities $\Gamma$ given $\pi$ and $(\Phi, \Theta)$ until the convergence of the log likelihood in equ.~\eqref{eq:logl}. The nested EM is reasonable, since the maximum a posteriori of the E-gene attachments $\Theta$ dependends on the weighted data $R_k$, which changes for every M step once we re-calculate the responsibilities $\Gamma$ in the E step.

We iterate between both steps until the mixture likelihood (equ.~\eqref{eq:logl}) converges.

\section{Simulations}
We show that M\&NEMs work in silico. For three, five and ten S-genes and $k=\left\{2,3,4,5\right\}$ we draw random mixture weights $\pi$ and components $\Phi, \Theta$. We simulate $1000$ cells according to $\pi$ and the components. The simulated data are log ratios with added Gaussian noise around $-1$ for no effect and $1$ for effect. Figure~\ref{fig:sim} shows the results for $100$ runs and Gaussian noise parameters $\sigma = \left\{0.1, 1, 2, 5\right\}$.

We compute accuracy by looking for similarity of rows of the adjacency matrices between the ground truth and inferred components. We cannot compare pairwise components due to model identifiability (supplement). We compute the accuracy of the mixture weights by the sorted euclidean distance between the ground truth and inferred mixture weights.

\begin{figure}
\includegraphics[width=\textwidth]{sim.pdf}
\caption{\textbf{Simulation results} The rows show results for $k=\left\{2,3,4,5\right\}$. The first column shows the running time for M\&NEM, NEM and random for different noise levels (x-axis) in seconds. The second column shows the inferred number of components. E.g. if M\&NEM infers 4 components, while the ground truth has 3, the value is $~1.33$. NEM and random by definition only infer one component. The third column shows the accuracy of the components. The fourth columns shows the distance between the estimated and ground truth sorted mixture weights.}\label{fig:sim}
\end{figure}

\section{Application to pooled single cell CrispR screens}

\subsection{Combining CrispR-based perturbation and RNA-seq (Perturb-Seq)}

The data set of Dixit, et al. 2016 consists of RNA-seq transcriptome read-outs for single cells. The data-set is available from the GEO database (GSE90063) or the single cell portal (\url{https://portals.broadinstitute.org/single_cell}). In the following analysis we used the log transformed transcripts per million normalised counts from the single cell portal. 

\begin{figure}
\includegraphics[width=1\textwidth]{perturbseq_ex.pdf}
\caption{\textbf{Perturb-seq result}}\label{fig:cropseq}
\end{figure}

\subsection{CrispR droplet sequencing (CROP-seq)}

Datlinger, et al. 2017 combine pooled CrispR screening with single-cell RNA sequencing to produce gene expression count data on the single cell level. They show the validity of their method with an analysis of T-cell receptor (TCR) activation in Jurkat cells. The processed CROP-seq data if available from the NCBI GEO database (GSE92872). We reduced the data to stimulated cells and genes, which have a median count number of $> 0$ over the remaining cells. We normalised the count data for library size, took the log of the normalised counts and calculated log likelihood ratios (supplement).

We map each gene to pathways from the KEGG (Kanehisa et al.) database. This leaves us with ten pathways with each containing a subset of the knock-down targets. For each pathway we take the subset of the data set including only those targets which are involved in the pathway. We use M\&NEM to infer the most likely mixture of pathways. We start with number of components $k=2$ and increase $k$ as long as the Bayesian information criteria increases. 

We show the results for all ten pathways in the supplement. Figure~\ref{fig:cropseq} shows the highest scoring (BIC) pathway. It is known that LCK is involved in the activation of ZAP70 (wikipedia). However, our two component mixture is evidence for a feedback loop, which is active in a subpopulation of cells and ZAP70 and LCK are causally influencing each other.

\begin{figure}
\includegraphics[width=1\textwidth]{cropseq_ex.pdf}
\caption{\textbf{Crop-seq result} Highest scoring pathway for the Crop-seq dataset. Circles show the perturbed genes. Boxes denote the a posteriori E-gene attachments. Diamonds show the numbers for the a posteriori hard clustering of the cells. Most cells agree with the known activation of ZAP70 by LCK \textbf{(A)}, however roughly one fourth show evidence for an activation of LCK by ZAP70 \textbf{(B)}.}\label{fig:cropseq}
\end{figure}

\section{Discussion}

We have introduced a novel method (M\&NEM) for the identification of heterogeneous sub populations of single cells with a mixture of networks. M\&NEM successfully infers sub populations and the underlaying ground truth mixture of networks in a simulation study under reasonable assumptions. In our application study we investigate two data sets from single cell CrispR experiments combined with full transcriptomic read-outs.

biology

M\&NEM is available s an R package at \url{https://bitbucket.org/MartinFXP/mnem/}.

\section{Supplement}

\subsection{Model identifiability}

In the case of the original NEMs, two networks $\Phi_1$ and $\Phi_2$ are identical if and only if they have equal transitive closures. We denote this with $\Phi_1 = \Phi_2$. This identity still holds for mixture NEMs. However, mixture NEMs yield additional identifiability problems. In special cases two different mixture $(\Phi^1_1, \Phi_2^1)$ and $(\Phi_1^2, \Phi_2^2)$ are identical even if $\Phi_i^j \neq \Phi_k^l ~\forall i \neq k \text{ or } j \neq l$.

Figure~\ref{fig:ident} shows an example of two mixtures with $k=2$ each. If we assume $\Theta$ is equal for all four $\Phi$, the two mixtures are identical, because they generate the same data. 

\begin{figure}
\begin{center}
\includegraphics[width=0.425\textwidth]{identnetsA.pdf}\hspace{2cm}\includegraphics[width=0.425\textwidth]{identnetsB.pdf}\\
\includegraphics[width=0.425\textwidth]{identmapsA.pdf}\hspace{2cm}\includegraphics[width=0.425\textwidth]{identmapsB.pdf}
\caption{\textbf{Model identifiability} \textbf{(A)} Mixture of two components (blue, red) with their respective attached cells (data) below. Dark areas are effects and light areas are no effects. Each column of the data is a cell and each row is the effect pattern for gene $E\_X$ attached to S-gene $X$. \textbf{(B)} Identical mixture to \textbf{A}. The first columns (cells) have been switched. Overall the data stays the same, but the network components changed.}\label{fig:ident}
\end{center}
\end{figure}

\subsection{Data processing}

\paragraph{Crop-seq}
We normalise the count values of each cell to its library size per million. Let $d_{ij}$ be the normalised count value for gene $i$ and cell $j$. Cell $j$  was perturbed by a knock-down of gene $k$. We estimate the density of the distribution of control counts for gene $i$ and the density for the distribution of counts from perturbed cells for gene $i$. We define the log likelihood ratio for $d_{ij}$ being affected by the knock-down by
\[r_{ij} = \mathrm{log} \left(\frac{p_{ij}^k}{p_{ij}^c}\right),\]
with $p_{ij}^c$ as the density value for $d_{ij}$ for the control distribution and $p_{ij}^k$ as the density value for the distribution of the knock-down of $k$. If the E-gene shows a clear effect in the cell $r_{ij}$ will be greater than zero and if it shows no effect it will be less than or equal to zero.

\section{References}
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4648179/

https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-017-0600-4

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3799884/

http://www.sciencedirect.com/science/article/pii/S0014579314001240

http://www.sciencedirect.com/science/article/pii/S2405471215000563?via\%3Dihub

\end{document}